### PR TITLE
Refactor micrometer configuration to use the Contributor mechanism

### LINKED ISSUE #161: Refactor micrometer-configuration to avoid reflection and use the Contributor API
We want to rely less on Reflection in case we want to use native compilation in future. Also the reflection loading is similar to the contributor mechanism so we can refactor to use it for Micrometer configuration.

### PR DESCRIPTION
**Pull request by robobario**
* Remove ability to load arbitrary micrometer config hook by Class name
* Remove ability to register arbitrary micrometer Meter Binders by classname
* Add ability to load user-contributed MicrometerConfigurationHook using a Contributor
* Refactor so common tags of the global registry can be configured by a default Contributor
* Refactor so standard Meter Binders can be registered with the global registry by a default Contributor

Standard Binders are configured like:
``` yaml
micrometer:
- type: "StandardBinders"
  config:
    binderNames:
    - "JvmGcMetrics"
```

Common Tags are configured like:
``` yaml
micrometer:
- type: "CommonTags"
  config:
    commonTags:
      a: "b"
```

There is an intention to keep Kroxylicious more native-compilation friendly so we should avoid Reflection where possible and use the JDK service loading mechanism to load user-contributed plugins from the classpath.

Closes #161

### CODE REVIEW

**Review by k-wall — APPROVED** [review]

File: integrationtests/src/test/java/io/kroxylicious/proxy/KroxyConfigBuilder.java [hunk 1]
```diff
@@ -80,25 +66,39 @@ public KroxyConfigBuilder withDefaultCluster(String bootstrapServers) {
     }
 
     public KroxyConfigBuilder withMicrometerBinder(String binder) {
-        MicrometerConfig original = currentOrNewMicrometerConfig();
-        micrometer = original.withBinder(binder);
+        MicrometerConfig config = getMicrometerConfigForType("StandardBinders");
+        @SuppressWarnings("unchecked")
+        List<String> binderNames = (List<String>) config.config.computeIfAbsent("binderNames", s -> new ArrayList<String>());
```
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/proxy/KroxyConfigBuilder.java:71]: `config.config` looks a bit funky.

File: integrationtests/src/test/java/io/kroxylicious/proxy/KroxyConfigBuilder.java [hunk 2]
```diff
@@ -80,25 +66,39 @@ public KroxyConfigBuilder withDefaultCluster(String bootstrapServers) {
...
         return this;
     }
 
-    public KroxyConfigBuilder withMicrometerConfigHook(String className) {
-        MicrometerConfig original = currentOrNewMicrometerConfig();
-        micrometer = original.withConfigHook(className);
+    public KroxyConfigBuilder withMicrometerCommonTag(String key, String value) {
+        MicrometerConfig config = getMicrometerConfigForType("CommonTags");
+        @SuppressWarnings("unchecked")
+        Map<String, String> commonTags = (Map<String, String>) config.config.computeIfAbsent("commonTags", s -> new HashMap<String, String>());
```
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/proxy/KroxyConfigBuilder.java:79]: nit: using `CommonTags` on line 77 and then `commonTags` here seems a bit inconsistent.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/micrometer/DefaultMicrometerConfigurationHookContributor.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.proxy.micrometer;
+
+import io.kroxylicious.proxy.service.BaseContributor;
+
+public class DefaultMicrometerConfigurationHookContributor extends BaseContributor<MicrometerConfigurationHook> implements MicrometerConfigurationHookContributor {
```
Comment (SamBarker) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/micrometer/DefaultMicrometerConfigurationHookContributor.java:10]: I'm not clear why we have `MicrometerConfigurationHookContributor` shouldn't `CommonTagsHook.CommonTagsHookConfig` offer its own Contributor and `StandardBindersHook.StandardBindersHookConfig` its own? I'm not seeing the value of the shared interface.

I could easily be missing something blindingly obvious...

**Review by SamBarker — APPROVED** [review]
I like the idea of switching from reflection to the service loader, not exactly clear on some of the relative merits of the specifics of the change.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/micrometer/DefaultMicrometerConfigurationHookContributor.java (hunk 3 above)
Comment (SamBarker) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/micrometer/DefaultMicrometerConfigurationHookContributor.java:10]: > I'm not clear why we have `MicrometerConfigurationHookContributor` shouldn't `CommonTagsHook.CommonTagsHookConfig` offer its own Contributor and `StandardBindersHook.StandardBindersHookConfig` its own? I'm not seeing the value of the shared interface.
> 
> I could easily be missing something blindingly obvious...

I guess what I was really saying is: what does the type field bring us in the model:

```yaml
micrometer:
- type: "StandardBinders"
  config:
    binderNames:
    - "JvmGcMetrics"
```
Compared with

```yaml
micrometer:
  "StandardBinders":
      binderNames:
      - "JvmGcMetrics"
  ```

Comment (robobario) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/micrometer/DefaultMicrometerConfigurationHookContributor.java:10]: The `type` string is used along with some Jackson resolution magic to obtain the type of the `config` field using the `Contributor`.

Comment (robobario) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/micrometer/DefaultMicrometerConfigurationHookContributor.java:10]: But yeah I wonder which is preferable:

a) One contributor like:
```
micrometer:
- type: "DefaultConfigHook"
  config:
    binderNames:
    - "JvmGcMetrics"
    commonTags:
       a: "b"
```
b) Using one contributor that offers two types
c) Using two contributor implementations that offer one type each

Comment (SamBarker) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/micrometer/DefaultMicrometerConfigurationHookContributor.java:10]: Let's stick as it's currently on this branch.

**Pull request merged**